core/async/src/test_loop.rs

```diff
@@ -0,0 +1,372 @@
+//! This is a framework to test async code in a way that is versatile, deterministic,
```

Comment:
(Overall file comment)
This is a lot of new infrastructure.
The code looks good, is very clean, has enough comments.
The file introduces 8 structs, 5 of them are public.

Excited to see, how this infrastructure will be used to simplify existing tests.